### PR TITLE
Add support for verifying digests to CLI verify commands

### PR DESCRIPTION
**Pull request by facutuesca**
#### Summary
This PR changes the `sigstore verify` CLI command so that it accepts digests as inputs (in addition to files).
Concretely:

```shell
# Before
sigstore verify identity ..... path/to/artifact.zip  # can only verify an artifact

# After
sigstore verify identity ..... path/to/artifact.zip  # can verify an artifact
sigstore verify identity ... --bundle b.sigstore.json sha256:aaabbbb.....  # can also verify an artifact's digest
```

This is useful in scenarios where the user doesn't want/need to download the artifact, but has access to its digest.

Related to https://github.com/sigstore/sigstore-conformance/issues/157

#### Release Note

* CLI: The `sigstore verify` command now supports verifying digests. This means
  that the user can now pass a digest like `sha256:aaaa....` instead of the
  path to an artifact, and `sigstore-python` will verify it as if it was the
  artifact with that digest.

#### Documentation

Example added to the README

cc @woodruffw

### CODE REVIEW

File: sigstore/hashes.py [hunk 1]
```diff
@@ -55,3 +55,9 @@ def _as_prehashed(self) -> Prehashed:
         if self.algorithm == HashAlgorithm.SHA2_256:
             return Prehashed(hashes.SHA256())
         raise Error(f"unknown hash algorithm: {self.algorithm}")
+
+    def __str__(self) -> str:
+        """
+        Returns a str representation of this `Hashed`.
+        """
+        return f"{self.algorithm.name}:{self.digest.hex()}"
```
Comment (woodruffw) [inline sigstore/hashes.py:63]: Small unit test for this, please 🙂

**Review by woodruffw — DISMISSED** [review]
LGTM, nice work @facutuesca! One small request for a unit test.

File: README.md [hunk 2]
```diff
@@ -421,6 +423,19 @@ $ python -m sigstore verify identity foo.txt bar.txt \
...
+
+`sigstore-python` supports verifying digests directly, without requiring the artifact to be
+present. The digest should be prefixed with the `sha256:` string:
+
+```console
+# looks for foo.txt.sigstore
+$ python -m sigstore verify identity sha256:ce8ab2822671752e201ea1e19e8c85e73d497e1c315bfd9c25f380b7625d1691 \
+    --cert-identity 'hamilcar@example.com' \
+    --cert-oidc-issuer 'https://github.com/login/oauth'
+    --bundle 'foo.txt.sigstore.json'
```
Comment (di) [inline README.md:436]: Either the comment or the `--bundle` value is wrong here I think?

File: sigstore/_cli.py [hunk 3]
```diff
@@ -149,12 +165,25 @@ def _add_shared_verify_input_options(group: argparse._ArgumentGroup) -> None:
         default=os.getenv("SIGSTORE_BUNDLE"),
         help=("The Sigstore bundle to verify with; not used with multiple inputs"),
     )
+
+    def file_or_digest(arg: str) -> Hashed | Path:
+        if arg.startswith("sha256:"):
+            digest = bytes.fromhex(arg[len("sha256:") :])
+            if len(digest) != 32:
+                raise ValueError()
```
Comment (di) [inline sigstore/_cli.py:173]: Where will this be caught? We might want to provide a more informative error message here.

File: README.md (hunk 2 above)
Comment (facutuesca) [reply README.md]: yup, comment was wrong, fixed!

File: sigstore/_cli.py (hunk 3 above)
Comment (facutuesca) [reply sigstore/_cli.py:173]: `argparse` [catches](https://docs.python.org/3/library/argparse.html#type) it:
> The argument to type can be any callable that accepts a single string. If the function raises [ArgumentTypeError](https://docs.python.org/3/library/argparse.html#argparse.ArgumentTypeError), [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError), or [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError), the exception is caught and a nicely formatted error message is displayed. No other exception types are handled.

The only problem is that it ignores the error message, so whatever message I put in `ValueError("mymsg")`, the shown error is always:
```
sigstore verify identity: error: argument FILE_OR_DIGEST: invalid file_or_digest value: 'sha256:3072b971b74902ac37ee452649efbaaeca6cd80c93cef5e27c1f491d7aa2ab443'
```

Comment (di) [reply sigstore/_cli.py:173]: Hmm, would be nice to have a more specific message but this should be fine!

File: sigstore/hashes.py (hunk 1 above)
Comment (facutuesca) [reply sigstore/hashes.py:63]: Done, see last commit

**Review by woodruffw — APPROVED** [review]

**Pull request merged**